File: app/src/main/java/org/radilabs/weather/weather/Daily.kt
package org.radilabs.weather.weather

const val COMPLETE_DAY_POINTS = 8

fun aggregateDaily(points: List<ForecastPoint>, timezoneOffsetSeconds: Int): List<DailySummary> {
    val groups = linkedMapOf<String, MutableList<ForecastPoint>>()
    for (point in points) {
        val key = localDateKey(point.atMs, timezoneOffsetSeconds)
        groups.getOrPut(key) { mutableListOf() }.add(point)
    }
    return groups.entries
        .sortedBy { it.key }
        .map { (localDate, dayPoints) ->
            DailySummary(
                localDate = localDate,
                highC = dayPoints.maxOf { it.temperatureC },
                lowC = dayPoints.minOf { it.temperatureC },
                condition = pickRepresentativeCondition(dayPoints.map { it.condition }),
                precipitation = summarizePrecipitation(dayPoints.map { it.precipitation }),
                pointCount = dayPoints.size,
                partial = dayPoints.size < COMPLETE_DAY_POINTS,
            )
        }
}

private fun summarizePrecipitation(items: List<Precipitation>): Precipitation {
    val probs = items.mapNotNull { it.probabilityPercent }
    val rains = items.mapNotNull { it.rainMm }
    val snows = items.mapNotNull { it.snowMm }
    return Precipitation(
        probabilityPercent = probs.maxOrNull(),
        rainMm = if (rains.isEmpty()) null else rains.sum(),
        snowMm = if (snows.isEmpty()) null else snows.sum(),
    )
}

File: src/weather/daily.ts
import { pickRepresentativeCondition } from "./conditions";
import type { DailySummary, ForecastPoint, Precipitation } from "./models";
import { localDateKey } from "./units";

/** A local calendar day is complete when it has a full 3-hour set (8 points). */
export const COMPLETE_DAY_POINTS = 8;

export function aggregateDaily(
  points: ForecastPoint[],
  timezoneOffsetSeconds: number,
): DailySummary[] {
  const groups = new Map<string, ForecastPoint[]>();
  for (const point of points) {
    const key = localDateKey(point.atMs, timezoneOffsetSeconds);
    const list = groups.get(key);
    if (list) list.push(point);
    else groups.set(key, [point]);
  }

  const days: DailySummary[] = [];
  for (const [localDate, dayPoints] of groups) {
    const temps = dayPoints.map((p) => p.temperatureC);
    days.push({
      localDate,
      highC: Math.max(...temps),
      lowC: Math.min(...temps),
      condition: pickRepresentativeCondition(dayPoints.map((p) => p.condition)),
      precipitation: summarizePrecipitation(dayPoints.map((p) => p.precipitation)),
      pointCount: dayPoints.length,
      partial: dayPoints.length < COMPLETE_DAY_POINTS,
    });
  }
  return days.sort((a, b) => a.localDate.localeCompare(b.localDate));
}

function summarizePrecipitation(items: Precipitation[]): Precipitation {
  const probs = items
    .map((p) => p.probabilityPercent)
    .filter((n): n is number => n !== undefined);
  const rains = items
    .map((p) => p.rainMm)
    .filter((n): n is number => n !== undefined);
  const snows = items
    .map((p) => p.snowMm)
    .filter((n): n is number => n !== undefined);

  const precipitation: Precipitation = {};
  if (probs.length > 0) precipitation.probabilityPercent = Math.max(...probs);
  if (rains.length > 0) precipitation.rainMm = rains.reduce((a, b) => a + b, 0);
  if (snows.length > 0) precipitation.snowMm = snows.reduce((a, b) => a + b, 0);
  return precipitation;
}
